### PR TITLE
[primer] Parallelize primer workflow into package batches

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description
The primer run now completes in [half the time](https://github.com/jacobtylerwalls/pylint/actions/runs/5003019925) (53m -> 28m) by dividing primer packages into batches.

The upside mainly consists of separating `home-assistant` from `pandas`, the two largest projects.

## Motivations
- Better development experience, shorter feedback loops
- Primer tool less likely to crash with out-of-memory issues trying to write huge .txt files (see https://github.com/pylint-dev/pylint/pull/8663#pullrequestreview-1427118576)
- Makes it more pleasant to debug multiprocessing issues (getting the primer to work with --jobs=0 would then provide more benefit)

## Notes
- We could shave a few more seconds off if `home-assistant` and `pandas` are given their own exclusive runs, but the few seconds it takes to lint another package in a batch after `pandas` is not substantial. My hope is that this is good enough and more maintainable.
-  Since the new arguments are optional, this is backwards-compatible with the existing way of running the primer locally without batching.

### CODE REVIEW

File: .github/workflows/primer_run_pr.yaml
Comment (DanielNoord) [inline .github/workflows/primer_run_pr.yaml:42]: Guess we can't do `len(batchIdx)` or something like that?

File: .github/workflows/primer_comment.yaml
Comment (DanielNoord) [inline .github/workflows/primer_comment.yaml:106]: Have you tested that the downloading of artifacts and unzipping works? As in, do the different batch jobs get zipped into one single file?

File: pylint/testutils/_primer/primer_run_command.py
Comment (DanielNoord) [inline pylint/testutils/_primer/primer_run_command.py:37]: Smart way to iterate this!

File: pylint/testutils/_primer/primer_run_command.py
Comment (DanielNoord) [inline pylint/testutils/_primer/primer_run_command.py:48]: Should we add a check here whether there are batches defined? Otherwise this name doesn't make a lot of sense.

**Review by DanielNoord — COMMENTED** [review]
Thanks for doing this! Very nice!

File: .github/workflows/primer_run_pr.yaml
Comment (jacobtylerwalls) [reply .github/workflows/primer_run_pr.yaml:42]: Yeah, unless we add a step to use bash or python to do math, I don't think I can do it inside a string formatting expression.

File: .github/workflows/primer_comment.yaml
Comment (jacobtylerwalls) [reply .github/workflows/primer_comment.yaml:106]: The 4 files get [zipped up](https://github.com/jacobtylerwalls/pylint/actions/runs/5003019925) into one .zip, and when you download it you can examine the four files.

File: pylint/testutils/_primer/primer_run_command.py
Comment (jacobtylerwalls) [reply pylint/testutils/_primer/primer_run_command.py:48]: Right, it will just say `_batchNone` as of now. I can clean this up a little.

Comment (jacobtylerwalls) [reply pylint/testutils/_primer/primer_run_command.py]: b94d0131

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks pretty good ! I think the python parts could be tested in ``tests/testutils/``. (This would be reused in a pylint-testutil package accessible to all pylint plugins maintainer).

**Comment by Pierre-Sassoulas** [comment]
I think we had issues with primers in the past that would not have happened if the python parts were covered from the get go (even if it's impossible to test the github actions code). That being said, the goal of this MR is not to add tests for everything that is not tested at the moment (only the new batch processing mechanism). I'd approve anyway if you don't want to invest the time this is clearly an improvement compared to what we have right now.

**Comment by jacobtylerwalls** [comment]
@Pierre-Sassoulas @DanielNoord Does anyone know why the `prettier` pre-commit hook is failing?

**Comment by Pierre-Sassoulas** [comment]
It seems as if prettier was outputting the path of every file in the repository without separators ?

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM. I fixed pre-commit but the tests now fails (?)

**Comment by jacobtylerwalls** [comment]
> I fixed pre-commit but the tests now fails (?)

Just looks like a transient failure from the Action we're using.

Thanks for the reviews!

**Pull request merged**
